analysis/topEFT/datacard_maker.py

```diff
@@ -372,11 +381,19 @@ def processSyst(process, systMap, d_hists, fout):
                     else:
                         systMap[syst] = {process: round(h_sys.Integral(), 3)}
             for syst_special,val in self.syst_special.items():
-                if syst_special not in process: continue
-                if syst in systMap:
-                    systMap[syst_special+'_flat_rate'].update({process: 1+val})
+                # Check for special bins
+                if isinstance(val,dict) and not any((b in fout.GetName() for b in val.keys())): continue
+                if isinstance(val,dict):
+                    for b in val:
+                        if b in fout.GetName(): # Might need to improve this: works for '2lss' not sure about '2lss_p'
+                            syst = val[b]
+                            break
+                else: syst = val
+                syst_cat = syst_special+'_flat_rate'
+                if any((syst_cat in s for s in systMap)):# and syst_special in systMap[syst_cat]:
+                    systMap[syst_special+'_flat_rate'].update({process: 1+syst})
                 else:
-                    systMap[syst_special+'_flat_rate'] = {process: 1+val}
+                    systMap[syst_cat] = {process: 1+syst}
```

Comment:
The `any` part is actually checking for times when a `systMap` entry is another dictionary. The keys will always be strings, but the values could be something like `{'2lss': 0.3}`. I'll try to clean up this line so it's clear it _only_ looks at the values and not the keys.
I see you also recommended we remove the `flat_rate` part, which I agree is probably better.

Reply:
I pushed some changes. Let me know if any of it still isn't clear.

Reply:
> The `any` part is actually checking for times when a `systMap` entry is another dictionary. The keys will always be strings, but the values could be something like `{'2lss': 0.3}`. I'll try to clean up this line so it's clear it _only_ looks at the values and not the keys. I see you also recommended we remove the `flat_rate` part, which I agree is probably better.

I'm confused here. The `any` keyword checks the boolean evaluation for each element of an iterable. In this case the iterable is the list of keys for the `systMap` dictionary and the boolean check is `(syst_cat in s)`. Now unless I'm mistaken, `s` will always be a string (or more specifically a _key_ of the dictionary), so I'm not sure how this checks for whether or not a given _value_ of the `systMap` dictionary will be another dictionary? In fact, it looks to me like regardless of the outcome of this check it is always assumed that the corresponding value of `systMap` will be a dictionary, since in the first case it calls `.update(...)` which is a method of a python dictionary, and in the other case it directly sets the value to a (newly initialized) dictionary.

Reply:
Alright thanks! I think this makes more sense now (from an outsider's perspective looking into the code).